File: internal/application/usecase/role/get.go
package role

import (
	"context"

	"go-clean-template/internal/domain"
)

type GetRoleInput struct {
	ID string
}

type GetRoleOutput struct {
	Role *domain.Role
}

type GetRoleUseCase interface {
	Execute(ctx context.Context, input *GetRoleInput) (*GetRoleOutput, error)
}

type getRoleUseCase struct {
	roleRepo domain.RoleRepository
}

func NewGetRoleUseCase(roleRepo domain.RoleRepository) GetRoleUseCase {
	return &getRoleUseCase{roleRepo: roleRepo}
}

func (uc *getRoleUseCase) Execute(ctx context.Context, input *GetRoleInput) (*GetRoleOutput, error) {
	role, err := uc.roleRepo.FindByID(ctx, input.ID)
	if err != nil {
		return nil, err
	}
	return &GetRoleOutput{Role: role}, nil
}

File: src/application/usecase/role/get.rs
use std::sync::Arc;

use async_trait::async_trait;

use crate::application::error::AppError;
use crate::application::usecase::user::RoleRepository;
use crate::domain;

pub struct GetRoleInput {
    pub id: String,
}

pub struct GetRoleOutput {
    pub role: domain::Role,
}

#[async_trait]
pub trait GetRoleUseCase: Send + Sync {
    async fn execute(&self, input: GetRoleInput) -> Result<GetRoleOutput, AppError>;
}

pub struct GetRoleUseCaseImpl {
    role_repo: Arc<dyn RoleRepository>,
}

impl GetRoleUseCaseImpl {
    pub fn new(role_repo: Arc<dyn RoleRepository>) -> Self {
        GetRoleUseCaseImpl { role_repo }
    }
}

#[async_trait]
impl GetRoleUseCase for GetRoleUseCaseImpl {
    async fn execute(&self, input: GetRoleInput) -> Result<GetRoleOutput, AppError> {
        let role = self.role_repo.find_by_id(&input.id).await?;
        Ok(GetRoleOutput { role })
    }
}
